File: src/zoho_books_cli/commands/customer_payments.py
"""`zb customer-payments ...` — full coverage of /customerpayments.

Thin wrappers: CRUD + update-by-custom-field + refunds sub-app. The singular
`/customerpayment/{id}/customfields` endpoint and the collection-level bulk
DELETE are left to `zb raw`.
"""

from __future__ import annotations

import typer

from zoho_books_cli import config
from zoho_books_cli.client import ZohoBooksClient
from zoho_books_cli.commands import _shared

app = typer.Typer(
    help="Customer payment operations (CRUD + update-by-custom-field + refunds).",
    no_args_is_help=True,
)
refunds_app = typer.Typer(help="Refunds on a customer payment (CRUD).", no_args_is_help=True)
app.add_typer(refunds_app, name="refunds")

BASE = "/customerpayments"


@app.command("list")
def list_payments(
    query: list[str] = typer.Option(
        None, "--query", "-q", help="Query params as key=value. May be repeated."
    ),
    params: str = typer.Option(
        None,
        "--params",
        help="Query params as a JSON object. Merged on top of --query.",
    ),
    page: int = typer.Option(None, "--page", help="Page number (1-indexed)."),
    per_page: int = typer.Option(None, "--per-page", help="Rows per page."),
):
    """List customer payments. Returns one page plus page_context."""
    q = _shared.parse_query_pairs(query, params)
    if page is not None:
        q["page"] = str(page)
    if per_page is not None:
        q["per_page"] = str(per_page)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.get(BASE, query=q)
    _shared.emit_list(resp, "customerpayments")


@app.command("create")
def create_payment(
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Create a customer payment."""
    json_body = _shared.parse_body(body)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(BASE, json_body=json_body)
    _shared.emit_object(resp)


@app.command("get")
def get_payment(
    payment_id: str = typer.Argument(..., help="Zoho Books customer_payment_id."),
):
    """Get a single customer payment by ID."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.get(f"{BASE}/{payment_id}")
    _shared.emit_object(resp)


@app.command("update")
def update_payment(
    payment_id: str = typer.Argument(..., help="Zoho Books customer_payment_id."),
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Update a customer payment by ID."""
    json_body = _shared.parse_body(body)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.put(f"{BASE}/{payment_id}", json_body=json_body)
    _shared.emit_object(resp)


@app.command("update-by-custom-field")
def update_payment_by_custom_field(
    key: str = typer.Option(..., "--key", help="Custom-field API name."),
    value: str = typer.Option(..., "--value", help="Custom-field value to match on."),
    body: str = typer.Option(
        ..., "--body", "-b", help="JSON body with the update fields. IDs must be strings."
    ),
    upsert: bool = typer.Option(
        False, "--upsert", help="Create a new customer payment if no match is found."
    ),
):
    """Update a customer payment by a custom field's unique value.

    Identifier is sent via X-Unique-Identifier-Key / X-Unique-Identifier-Value headers
    per Zoho's spec; the request body carries only the fields to update.
    """
    json_body = _shared.parse_body(body)
    headers = {"X-Unique-Identifier-Key": key, "X-Unique-Identifier-Value": value}
    if upsert:
        headers["X-Upsert"] = "true"
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.put(BASE, json_body=json_body, headers=headers)
    _shared.emit_object(resp)


@app.command("delete")
def delete_payment(
    payment_id: str = typer.Argument(..., help="Zoho Books customer_payment_id."),
):
    """Delete a customer payment by ID."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.delete(f"{BASE}/{payment_id}")
    _shared.emit_action("payment_id", payment_id, resp)


@refunds_app.command("list")
def list_refunds(
    payment_id: str = typer.Argument(..., help="Zoho Books customer_payment_id."),
    query: list[str] = typer.Option(None, "--query", "-q", help="Query params as key=value."),
    params: str = typer.Option(
        None,
        "--params",
        help="Query params as a JSON object. Merged on top of --query.",
    ),
    page: int = typer.Option(None, "--page", help="Page number (1-indexed)."),
    per_page: int = typer.Option(None, "--per-page", help="Rows per page."),
):
    """List refunds issued against a customer payment."""
    q = _shared.parse_query_pairs(query, params)
    if page is not None:
        q["page"] = str(page)
    if per_page is not None:
        q["per_page"] = str(per_page)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.get(f"{BASE}/{payment_id}/refunds", query=q)
    _shared.emit_list(resp, "payment_refunds")


@refunds_app.command("create")
def create_refund(
    payment_id: str = typer.Argument(..., help="Zoho Books customer_payment_id."),
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Issue a refund against a customer payment."""
    json_body = _shared.parse_body(body)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(f"{BASE}/{payment_id}/refunds", json_body=json_body)
    _shared.emit_object(resp)


@refunds_app.command("get")
def get_refund(
    payment_id: str = typer.Argument(..., help="Zoho Books customer_payment_id."),
    refund_id: str = typer.Argument(..., help="Zoho Books refund_id."),
):
    """Get a single refund on a customer payment."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.get(f"{BASE}/{payment_id}/refunds/{refund_id}")
    _shared.emit_object(resp)


@refunds_app.command("update")
def update_refund(
    payment_id: str = typer.Argument(..., help="Zoho Books customer_payment_id."),
    refund_id: str = typer.Argument(..., help="Zoho Books refund_id."),
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Update a refund on a customer payment."""
    json_body = _shared.parse_body(body)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.put(f"{BASE}/{payment_id}/refunds/{refund_id}", json_body=json_body)
    _shared.emit_object(resp)


@refunds_app.command("delete")
def delete_refund(
    payment_id: str = typer.Argument(..., help="Zoho Books customer_payment_id."),
    refund_id: str = typer.Argument(..., help="Zoho Books refund_id."),
):
    """Delete a refund on a customer payment."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.delete(f"{BASE}/{payment_id}/refunds/{refund_id}")
    _shared.emit_action("refund_id", refund_id, resp)

File: src/commands/customer_payments.rs
//! `zb customer-payments` — CRUD + update-by-custom-field + refunds sub-app.

use clap::{Args, Subcommand};

use crate::cli::Ctx;
use crate::commands::common::{self, BodyArgs, CustomFieldUpdateArgs, ListArgs};
use crate::errors::Result;

const BASE: &str = "/customerpayments";

#[derive(Args, Debug)]
pub struct Cmd {
    #[command(subcommand)]
    pub sub: Sub,
}

#[derive(Subcommand, Debug)]
pub enum Sub {
    /// List customer payments.
    List(ListArgs),
    /// Create a customer payment.
    Create(BodyArgs),
    /// Get a single customer payment by ID.
    Get(IdArgs),
    /// Update a customer payment by ID.
    Update(UpdateArgs),
    /// Update a customer payment by a custom field's unique value.
    #[command(name = "update-by-custom-field")]
    UpdateByCustomField(CustomFieldUpdateArgs),
    /// Delete a customer payment by ID.
    Delete(IdArgs),
    /// Refunds on a customer payment.
    Refunds(RefundsCmd),
}

#[derive(Args, Debug)]
pub struct IdArgs {
    pub payment_id: String,
}

#[derive(Args, Debug)]
pub struct UpdateArgs {
    pub payment_id: String,
    #[command(flatten)]
    pub body: BodyArgs,
}

#[derive(Args, Debug)]
pub struct RefundsCmd {
    #[command(subcommand)]
    pub sub: RefundsSub,
}

#[derive(Subcommand, Debug)]
pub enum RefundsSub {
    /// List refunds issued against a customer payment.
    List(RefundsListArgs),
    /// Issue a refund against a customer payment.
    Create(RefundsCreateArgs),
    /// Get a single refund on a customer payment.
    Get(RefundIdArgs),
    /// Update a refund on a customer payment.
    Update(RefundsUpdateArgs),
    /// Delete a refund on a customer payment.
    Delete(RefundIdArgs),
}

#[derive(Args, Debug)]
pub struct RefundsListArgs {
    pub payment_id: String,
    #[command(flatten)]
    pub list: ListArgs,
}

#[derive(Args, Debug)]
pub struct RefundsCreateArgs {
    pub payment_id: String,
    #[command(flatten)]
    pub body: BodyArgs,
}

#[derive(Args, Debug)]
pub struct RefundIdArgs {
    pub payment_id: String,
    pub refund_id: String,
}

#[derive(Args, Debug)]
pub struct RefundsUpdateArgs {
    pub payment_id: String,
    pub refund_id: String,
    #[command(flatten)]
    pub body: BodyArgs,
}

pub fn run(cmd: Cmd, ctx: &mut Ctx) -> Result<()> {
    match cmd.sub {
        Sub::List(args) => common::list(ctx, BASE, &args, "customerpayments"),
        Sub::Create(args) => common::create(ctx, BASE, &args),
        Sub::Get(args) => common::get(ctx, &format!("{BASE}/{}", args.payment_id)),
        Sub::Update(args) => {
            common::update(ctx, &format!("{BASE}/{}", args.payment_id), &args.body)
        }
        Sub::UpdateByCustomField(args) => common::update_custom(ctx, BASE, &args),
        Sub::Delete(args) => {
            let path = format!("{BASE}/{}", args.payment_id);
            common::delete(ctx, &path, "payment_id", &args.payment_id)
        }
        Sub::Refunds(r) => match r.sub {
            RefundsSub::List(args) => common::nested_list(
                ctx,
                BASE,
                &args.payment_id,
                "refunds",
                &args.list,
                "payment_refunds",
            ),
            RefundsSub::Create(args) => {
                let path = format!("{BASE}/{}/refunds", args.payment_id);
                common::create(ctx, &path, &args.body)
            }
            RefundsSub::Get(args) => {
                let path = format!("{BASE}/{}/refunds/{}", args.payment_id, args.refund_id);
                common::get(ctx, &path)
            }
            RefundsSub::Update(args) => {
                let path = format!("{BASE}/{}/refunds/{}", args.payment_id, args.refund_id);
                common::update(ctx, &path, &args.body)
            }
            RefundsSub::Delete(args) => {
                let path = format!("{BASE}/{}/refunds/{}", args.payment_id, args.refund_id);
                common::delete(ctx, &path, "refund_id", &args.refund_id)
            }
        },
    }
}
